pkg/apiclient/grpcproxy.go

```diff
@@ -0,0 +1,202 @@
+package apiclient
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	argocderrors "github.com/argoproj/argo-cd/errors"
+	"github.com/argoproj/argo-cd/util"
+)
+
+const (
+	frameHeaderLength = 5
+	endOfStreamFlag   = 128
+)
+
+type noopCodec struct{}
+
+func (noopCodec) Marshal(v interface{}) ([]byte, error) {
+	return v.([]byte), nil
+}
+
+func (noopCodec) Unmarshal(data []byte, v interface{}) error {
+	pointer := v.(*[]byte)
+	*pointer = data
+	return nil
+}
+
+func (noopCodec) String() string {
+	return "bytes"
+}
+
+type inlineCloser struct {
+	close func() error
+}
+
+func (c *inlineCloser) Close() error {
+	return c.close()
+}
+
+func toFrame(msg []byte) []byte {
+	frame := append([]byte{0, 0, 0, 0}, msg...)
+	binary.BigEndian.PutUint32(frame, uint32(len(msg)))
+	frame = append([]byte{0}, frame...)
+	return frame
+}
+
+func (c *client) executeRequest(fullMethodName string, msg []byte, md metadata.MD) (*http.Response, error) {
+	schema := "https"
+	if c.PlainText {
+		schema = "http"
+	}
+	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("%s://%s%s", schema, c.ServerAddr, fullMethodName), bytes.NewReader(toFrame(msg)))
+	if err != nil {
+		return nil, err
+	}
+	if md != nil {
+		for k, v := range md {
+			if strings.HasPrefix(k, ":") {
+				continue
+			}
+			for i := range v {
+				req.Header.Set(k, v[i])
+			}
+		}
+	}
+	req.Header.Set("content-type", "application/grpc-web+proto")
+
+	client := &http.Client{Transport: &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: c.Insecure},
+	}}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	var code codes.Code
+	if statusStr := resp.Header.Get("Grpc-Status"); statusStr != "" {
+		statusInt, err := strconv.Atoi(statusStr)
+		if err != nil {
+			code = codes.Unknown
+		} else {
+			code = codes.Code(statusInt)
+		}
+		if code != codes.OK {
+			return nil, status.Error(code, resp.Header.Get("Grpc-Message"))
+		}
+	}
+	return resp, nil
+}
+
+func (c *client) startGRPCProxy() (*grpc.Server, net.Listener, error) {
+	serverAddr := fmt.Sprintf("%s/argocd-%d.sock", os.TempDir(), time.Now().Unix())
```

Comment:
It looks like we are not cleaning this file up at all. I think this will cause a lot of litter if someone is scripting using argocd cli.

Reply:
The socket file is deleted when listener gets closed. Just in case tested it manually and file was successfully deleted